packages/Search/test/DTK_BoostRTreeHelpers.hpp

```diff
@@ -53,6 +58,83 @@ static RTree<typename View::value_type> makeRTree( View const &objects )
         boost::adaptors::transformed( PairMaker<Indexable, int>() ) );
 }
 
+// NOTE: Boost.Config defines BOOST_NO_CXX11_VARIADIC_TEMPLATES for nvcc
+// with the current version of CUDA we are using.  In consequence we are
+// not able to use std::tuple<Indexable, ...> which is unfortunate :(
+class AppendRankToPairObjectIndex
+{
+  public:
+    AppendRankToPairObjectIndex( int rank )
+        : _rank( rank )
+    {
+    }
+
+    template <typename T1, typename T2>
+    inline boost::tuple<T1, T2, int>
+    operator()( std::pair<T1, T2> const &p ) const
+    {
+        return boost::make_tuple( std::get<0>( p ), std::get<1>( p ), _rank );
+    }
+
+  private:
+    int _rank = -1;
+};
+
+template <typename Indexable>
+using ParallelRTree =
+    boost::geometry::index::rtree<boost::tuple<Indexable, int, int>, Parameter>;
+
+template <typename View>
+static ParallelRTree<typename View::value_type> makeRTree( MPI_Comm comm,
+                                                           View const &objects )
+{
+    using Indexable = typename View::value_type;
+    auto const n = objects.extent_int( 0 );
+
+    // Fill buffer with pair (object, index)
+    std::vector<std::pair<Indexable, int>> buffer;
+    boost::copy(
+        objects | boost::adaptors::indexed() |
+            boost::adaptors::transformed( PairMaker<Indexable, int>() ),
+        std::back_inserter( buffer ) );
+
+    // Gather all buffers
+    int comm_size;
+    MPI_Comm_size( comm, &comm_size );
+    int comm_rank;
+    MPI_Comm_rank( comm, &comm_rank );
+    std::vector<int> counts( comm_size );
+    counts[comm_rank] = buffer.size();
+    MPI_Allgather( MPI_IN_PLACE, 0, MPI_DATATYPE_NULL, counts.data(), 1,
+                   MPI_INT, comm );
+    std::vector<int> offsets( comm_size + 1 );
+    offsets[0] = 0;
+    for ( int i = 0; i < comm_size; ++i )
+        offsets[i + 1] = counts[i] + offsets[i];
+    decltype( buffer ) all_buffers( offsets.back() );
+    auto const bytes_per_element =
+        sizeof( typename decltype( buffer )::value_type );
+    for ( auto pv : {&counts, &offsets} )
+        for ( auto &x : *pv )
+            x *= bytes_per_element;
+    MPI_Allgatherv( buffer.data(), counts[comm_rank], MPI_BYTE,
+                    all_buffers.data(), counts.data(), offsets.data(), MPI_BYTE,
+                    comm );
+    for ( auto &x : offsets )
+        x /= bytes_per_element;
+
+    // Flatten the data structure and turn pair into tuple (object, indice,
+    // rank)
```

Comment:
Comment is outdated and dates back from when I was using Boost.MPI and `all_buffers` was a vector of vector.  Will remove.